### PR TITLE
Pr iast 2634 add iast integration with gitlab and GitHub

### PR DESCRIPTION
**Pull request by kecven**
By submitting a PR to this repository, you agree to the terms within the [Checkmarx Code of Conduct](https://github.com/checkmarx-ltd/open-source-template/blob/master/CODE-OF-CONDUCT.md). Please see the [contributing guidelines](https://github.com/checkmarx-ltd/open-source-template/blob/master/CONTRIBUTING.md) for how to create and submit a high-quality PR for this repo.

### Description

Add integration IAST and CX-flow (github / gitlab).

### References

> Include supporting link to GitHub Issue/PR number

### Testing

src/test/resources/cucumber/features/integrationTests/cli/iastCli.feature

### Checklist

- [ ] I have added documentation for new/changed functionality in this PR (if applicable).  *If documentation is a Wiki Update, please indicate desired changes within PR MD Comment*
- [ ] All active GitHub checks for tests, formatting, and security are passing
- [ ] The correct base branch is being used

### CODE REVIEW

**Comment by cxflowtestuser** [comment]
Scan submitted to Checkmarx

**Comment by cxflowtestuser** [comment]
Logo<br />**Checkmarx SAST - Scan Summary & Details**
### Cx-SAST Summary
Total of 1412 vulnerabilities<br />High&nbsp;**0 High**<br />Medium&nbsp;**3 Medium**<br />Low&nbsp;**1282 Low**<br />Info&nbsp;**127 Info**<br />
### Violation Summary
Medium&nbsp;**1** **Medium**<br />
[View more details on Checkmarx UI](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1067025&projectid=3396)
### Cx-SAST Details
|Lines|Severity|Category|File|Link|
---|---|---|---|---
[342](https://github.com/checkmarx-ltd/cx-flow/blob/pr-iast-2634-add_iast_integration_with_gitlab_and_github/src/main/java/com/checkmarx/flow/custom/GitLabIssueTracker.java#L342) |Medium|Privacy_Violation|src/main/java/com/checkmarx/flow/custom/GitLabIssueTracker.java|[Checkmarx](https://sast.checkmarx.net/CxWebClient/ViewerMain.aspx?scanid=1067025&projectid=3396&pathid=44)
***
Logo
**Checkmarx SCA - Scan Summary & Details**
### Cx-SCA Summary
**Total Packages Identified**: **269**
**Scan Risk Score**: **9.80**

High&nbsp;**11** **High** **severity vulnerabilities**
Medium&nbsp;**2** **Medium** **severity vulnerabilities**
Low&nbsp;**1** **Low** **severity vulnerabilities**
[View more details on Checkmarx UI](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560)
### Cx-SCA vulnerability result overview
|Vulnerability ID|Package|Severity|CVSS score|Publish date|Current version|Recommended version|Link in CxSCA|Reference – NVD link|
---|---|---|---|---|---|---|---|---
`CVE-2016-1000031`|commons-fileupload:commons-fileupload|HIGH|9.8|2016-10-25T14:29:00|1.0|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2016-1000031%3AMaven-commons-fileupload%3Acommons-fileupload-1.0/vulnerabilityDetails)|[CVE-2016-1000031](https://nvd.nist.gov/vuln/detail/CVE-2016-1000031)
`Cxa9261daf-3755`|org.bouncycastle:bcprov-jdk15on|HIGH|9.8|2021-04-06T11:30:00|1.68|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/Cxa9261daf-3755%3AGradle-org.bouncycastle%3Abcprov-jdk15on-1.68/vulnerabilityDetails)|N\A
`CVE-2016-1000027`|org.springframework:spring-web|HIGH|9.8|2020-01-02T23:15:00|5.2.10.RELEASE|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2016-1000027%3AGradle-org.springframework%3Aspring-web-5.2.10.RELEASE/vulnerabilityDetails)|[CVE-2016-1000027](https://nvd.nist.gov/vuln/detail/CVE-2016-1000027)
`CVE-2016-1000027`|org.springframework:spring-webmvc|HIGH|9.8|2020-01-02T23:15:00|5.2.10.RELEASE|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2016-1000027%3AGradle-org.springframework%3Aspring-webmvc-5.2.10.RELEASE/vulnerabilityDetails)|[CVE-2016-1000027](https://nvd.nist.gov/vuln/detail/CVE-2016-1000027)
`CVE-2021-22118`|org.springframework:spring-web|HIGH|7.8|2021-05-27T15:15:00|5.2.10.RELEASE|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2021-22118%3AGradle-org.springframework%3Aspring-web-5.2.10.RELEASE/vulnerabilityDetails)|[CVE-2021-22118](https://nvd.nist.gov/vuln/detail/CVE-2021-22118)
`CVE-2016-3092`|commons-fileupload:commons-fileupload|HIGH|7.5|2016-07-04T22:59:00|1.0|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2016-3092%3AMaven-commons-fileupload%3Acommons-fileupload-1.0/vulnerabilityDetails)|[CVE-2016-3092](https://nvd.nist.gov/vuln/detail/CVE-2016-3092)
`CVE-2013-2186`|commons-fileupload:commons-fileupload|HIGH|7.5|2013-10-28T21:55:00|1.0|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2013-2186%3AMaven-commons-fileupload%3Acommons-fileupload-1.0/vulnerabilityDetails)|[CVE-2013-2186](https://nvd.nist.gov/vuln/detail/CVE-2013-2186)
`CVE-2014-0050`|commons-fileupload:commons-fileupload|HIGH|7.5|2014-04-01T06:27:00|1.0|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2014-0050%3AMaven-commons-fileupload%3Acommons-fileupload-1.0/vulnerabilityDetails)|[CVE-2014-0050](https://nvd.nist.gov/vuln/detail/CVE-2014-0050)
`CVE-2020-17527`|org.apache.tomcat.embed:tomcat-embed-core|HIGH|7.5|2020-12-03T19:15:00|9.0.39|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2020-17527%3AGradle-org.apache.tomcat.embed%3Atomcat-embed-core-9.0.39/vulnerabilityDetails)|[CVE-2020-17527](https://nvd.nist.gov/vuln/detail/CVE-2020-17527)
`CVE-2021-25122`|org.apache.tomcat.embed:tomcat-embed-core|HIGH|7.5|2021-03-01T12:15:00|9.0.39|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2021-25122%3AGradle-org.apache.tomcat.embed%3Atomcat-embed-core-9.0.39/vulnerabilityDetails)|[CVE-2021-25122](https://nvd.nist.gov/vuln/detail/CVE-2021-25122)
`CVE-2021-25329`|org.apache.tomcat.embed:tomcat-embed-core|HIGH|7.0|2021-03-01T12:15:00|9.0.39|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2021-25329%3AGradle-org.apache.tomcat.embed%3Atomcat-embed-core-9.0.39/vulnerabilityDetails)|[CVE-2021-25329](https://nvd.nist.gov/vuln/detail/CVE-2021-25329)
`CVE-2021-24122`|org.apache.tomcat.embed:tomcat-embed-core|MEDIUM|5.9|2021-01-14T15:15:00|9.0.39|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2021-24122%3AGradle-org.apache.tomcat.embed%3Atomcat-embed-core-9.0.39/vulnerabilityDetails)|[CVE-2021-24122](https://nvd.nist.gov/vuln/detail/CVE-2021-24122)
`CVE-2020-17521`|org.codehaus.groovy:groovy|MEDIUM|5.5|2020-12-07T20:15:00|2.5.13|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2020-17521%3AGradle-org.codehaus.groovy%3Agroovy-2.5.13/vulnerabilityDetails)|[CVE-2020-17521](https://nvd.nist.gov/vuln/detail/CVE-2020-17521)
`CVE-2013-0248`|commons-fileupload:commons-fileupload|LOW|3.3|2013-03-15T20:55:00|1.0|| [Vulnerability Link](https://eu.sca.checkmarx.net/#/projects/716d8e63-ffc4-493d-8c95-331946293f0a/reports/7326c312-5df1-444a-85ed-1b5a194e6560/vulnerabilities/CVE-2013-0248%3AMaven-commons-fileupload%3Acommons-fileupload-1.0/vulnerabilityDetails)|[CVE-2013-0248](https://nvd.nist.gov/vuln/detail/CVE-2013-0248)

File: src/main/java/com/checkmarx/flow/CxFlowApplication.java [hunk 1]
```diff
@@ -11,6 +12,7 @@
 @SpringBootApplication(scanBasePackages = { "com.checkmarx.sdk", "com.checkmarx.flow", "com.cx.restclient"})
 @EnableAspectJAutoProxy(proxyTargetClass = true)
 @EnableAsync
+@EnableCaching
```
Comment (AvivCx) [inline src/main/java/com/checkmarx/flow/CxFlowApplication.java:15]: what is the purpose for this change?

File: src/main/java/com/checkmarx/flow/config/GitHubProperties.java [hunk 2]
```diff
@@ -32,9 +32,16 @@ public String getMergeNoteUri(String namespace, String repo, String mergeId){
...
     }
 
-    public String getGitUri(String namespace, String repo){
+    public String getGitUri(String namespace, String repo) {
         String format = "%s/%s/%s.git";
         return String.format(format, getUrl(), namespace, repo);
         //sample: https://github.com/namespace/repo.git
     }
+
+    public String getIssueUri(String namespace, String repo) {
```
Comment (AvivCx) [inline src/main/java/com/checkmarx/flow/config/GitHubProperties.java:41]: what is the purpose of this new method. where it is used?

File: src/main/java/com/checkmarx/flow/dto/iast/manager/dto/description/VulnerabilityDescription.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
+package com.checkmarx.flow.dto.iast.manager.dto.description;
```
Comment (AvivCx) [inline src/main/java/com/checkmarx/flow/dto/iast/manager/dto/description/VulnerabilityDescription.java:1]: why you need new vulnerability description?

File: src/main/java/com/checkmarx/flow/service/GitHubService.java [hunk 4]
```diff
@@ -8,12 +8,7 @@
...
 import com.checkmarx.flow.config.ScmConfigOverrider;
 import com.checkmarx.flow.constants.FlowConstants;
-import com.checkmarx.flow.dto.OperationResult;
-import com.checkmarx.flow.dto.RepoComment;
-import com.checkmarx.flow.dto.RepoIssue;
-import com.checkmarx.flow.dto.ScanDetails;
-import com.checkmarx.flow.dto.ScanRequest;
-import com.checkmarx.flow.dto.Sources;
+import com.checkmarx.flow.dto.*;
 import com.checkmarx.flow.dto.github.Content;
```
Comment (AvivCx) [inline src/main/java/com/checkmarx/flow/service/GitHubService.java:12]: What are the changes in this file? not clear

File: src/main/java/com/checkmarx/flow/service/GitLabService.java [hunk 5]
```diff
@@ -12,8 +12,8 @@
 import com.checkmarx.flow.exception.GitLabClientException;
 import com.checkmarx.flow.utils.HTMLHelper;
 import com.checkmarx.flow.utils.ScanUtils;
-import com.checkmarx.sdk.dto.sast.CxConfig;
 import com.checkmarx.sdk.dto.ScanResults;
+import com.checkmarx.sdk.dto.sast.CxConfig;
 import org.apache.commons.codec.binary.Base64;
 import org.json.JSONArray;
 import org.json.JSONException;
```
Comment (AvivCx) [inline src/main/java/com/checkmarx/flow/service/GitLabService.java:19]: What are the changes in this file? not clear

File: src/main/java/com/checkmarx/flow/service/RepoService.java [hunk 6]
```diff
@@ -37,17 +37,16 @@ public void sendMergeComment(ScanRequest request, String comment){
                     return;
                 }
                 log.debug("sendMergeComment: Going to update {} pull request comment",
-                          request.getRepoType());
+                        request.getRepoType());
                 updateComment(commentToUpdate.getCommentUrl(), comment, request);
             } else {
```
Comment (AvivCx) [inline src/main/java/com/checkmarx/flow/service/RepoService.java:42]: What are the changes in this file? not clear

File: src/main/java/com/checkmarx/flow/CxFlowApplication.java (hunk 1 above)
Comment (kecven) [reply src/main/java/com/checkmarx/flow/CxFlowApplication.java:15]: I used cache. I don't wanna to send many requests to gitlab. Check where I use anatation Cacheable

File: src/main/java/com/checkmarx/flow/config/GitHubProperties.java (hunk 2 above)
Comment (kecven) [reply src/main/java/com/checkmarx/flow/config/GitHubProperties.java:41]: yes. now don't use. will remove that

File: src/main/java/com/checkmarx/flow/dto/iast/manager/dto/description/VulnerabilityDescription.java (hunk 3 above)
Comment (kecven) [reply src/main/java/com/checkmarx/flow/dto/iast/manager/dto/description/VulnerabilityDescription.java:1]: I used that information from Iast. Don't understand a question. that is a DTO object which I get from IAST.

File: src/main/java/com/checkmarx/flow/service/GitHubService.java (hunk 4 above)
Comment (kecven) [reply src/main/java/com/checkmarx/flow/service/GitHubService.java:12]: optimization from IDEA

File: src/main/java/com/checkmarx/flow/service/GitLabService.java (hunk 5 above)
Comment (kecven) [reply src/main/java/com/checkmarx/flow/service/GitLabService.java:19]: optimization from IDEA

File: src/main/java/com/checkmarx/flow/service/RepoService.java (hunk 6 above)
Comment (kecven) [reply src/main/java/com/checkmarx/flow/service/RepoService.java:42]: optimization from IDEA

File: src/main/java/com/checkmarx/flow/CxFlowApplication.java (hunk 1 above)
Comment (AvivCx) [reply src/main/java/com/checkmarx/flow/CxFlowApplication.java:15]: I'm not sure what would be the impact of that. and I don't know how well it was tested
Can we remove this change from the PR?

**Comment by cxflowtestuser** [comment]
Scan not submitted to Checkmarx due to existing Active scan for the same project.

**Review by AvivCx — APPROVED** [review]

**Pull request merged**